File: frontend/src/app/contexts/journal/domain/weight.ts
export type WeightValidationError =
  { kind: 'number' } | { kind: 'positive' } | { kind: 'multiple-of-grams-unit'; gramsUnit: number };

export class Weight {
  private static readonly MEASUREMENT_INCREMENT_IN_GRAMS = 50;

  public static measurementIncrementInKilograms(): number {
    return Weight.kilogramsFromGrams(Weight.MEASUREMENT_INCREMENT_IN_GRAMS);
  }

  private static gramsFromKilograms(kilograms: number): number {
    const grams = kilograms * 1000;
    const nearestWholeGram = Math.round(grams);

    return nearestWholeGram / 1000 === kilograms ? nearestWholeGram : grams;
  }

  private static kilogramsFromGrams(grams: number): number {
    return grams / 1000;
  }

  private static validateWeightInGrams(grams: number): WeightValidationError[] {
    const errors: WeightValidationError[] = [];
    if (!Number.isSafeInteger(grams)) {
      errors.push({ kind: 'number' });
    }
    if (grams <= 0) {
      errors.push({ kind: 'positive' });
    }
    if (grams % Weight.MEASUREMENT_INCREMENT_IN_GRAMS !== 0) {
      errors.push({
        kind: 'multiple-of-grams-unit',
        gramsUnit: Weight.MEASUREMENT_INCREMENT_IN_GRAMS,
      });
    }
    return errors;
  }

  public static validateWeightInKilograms(kilograms: number): WeightValidationError[] {
    return Weight.validateWeightInGrams(Weight.gramsFromKilograms(kilograms));
  }

  private constructor(private readonly grams: number) {
    if (Weight.validateWeightInGrams(grams).length > 0) {
      throw new RangeError(
        `"${grams}" is not a valid weight in grams: must be a positive integer multiple of ${Weight.MEASUREMENT_INCREMENT_IN_GRAMS}`,
      );
    }
  }

  public static of(kilograms: number): Weight {
    const grams = Weight.gramsFromKilograms(kilograms);
    return new Weight(grams);
  }

  public inKilograms(): number {
    return this.grams / 1000;
  }
}

File: backend/src/main/java/fr/sylvainjanet/tracker/journal/domain/Weight.java
package fr.sylvainjanet.tracker.journal.domain;

import fr.sylvainjanet.tracker.journal.domain.error.WeightValidationError;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Objects;
import java.util.Set;

public final class Weight {

    private static final long GRAMS_MEASURABLE_UNIT = 50L;

    private final long grams;

    private Weight(BigDecimal grams) {
        validateWeightInGramsOrThrow(grams);
        this.grams = grams.longValueExact();
    }

    private static void validateWeightInGramsOrThrow(BigDecimal grams) {
        Set<WeightValidationError> errors = validateWeightInGrams(grams);
        if (!errors.isEmpty()) {
            throw new IllegalArgumentException(
                    "weight must be a positive number of grams that is a multiple of "
                            + GRAMS_MEASURABLE_UNIT
                            + " grams");
        }
    }

    private static Set<WeightValidationError> validateWeightInGrams(BigDecimal grams) {
        Set<WeightValidationError> errors = new java.util.HashSet<>();
        try {
            long gramsLong = grams.longValueExact();
            if (gramsLong <= 0) {
                errors.add(WeightValidationError.positive());
            }
            if (gramsLong % GRAMS_MEASURABLE_UNIT != 0) {
                errors.add(WeightValidationError.multipleOfGramsUnit((int) GRAMS_MEASURABLE_UNIT));
            }
        } catch (ArithmeticException _) {
            errors.add(WeightValidationError.multipleOfGramsUnit((int) GRAMS_MEASURABLE_UNIT));
        }
        return Set.copyOf(errors);
    }

    private static BigDecimal toKilogramsUnchecked(BigDecimal grams) {
        return BigDecimal.valueOf(grams.longValueExact())
                .divide(BigDecimal.valueOf(1000), 2, RoundingMode.UNNECESSARY);
    }

    private static BigDecimal toGramsUnchecked(BigDecimal kilograms) {
        return kilograms.multiply(BigDecimal.valueOf(1000));
    }

    public static BigDecimal toKilograms(BigDecimal grams) {
        validateWeightInGramsOrThrow(grams);
        return toKilogramsUnchecked(grams);
    }

    public static BigDecimal toGrams(BigDecimal kilograms) {
        BigDecimal grams = toGramsUnchecked(kilograms);
        validateWeightInGramsOrThrow(grams);
        return grams.setScale(0, RoundingMode.UNNECESSARY);
    }

    public static Set<WeightValidationError> validateWeightInKg(BigDecimal kilograms) {
        BigDecimal grams = toGramsUnchecked(kilograms);
        return validateWeightInGrams(grams);
    }

    public static Weight of(BigDecimal kilograms) {
        BigDecimal grams = toGrams(kilograms);
        return new Weight(grams);
    }

    public BigDecimal inKilograms() {
        return toKilograms(BigDecimal.valueOf(grams));
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        Weight weight = (Weight) o;
        return grams == weight.grams;
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(grams);
    }
}
